[서지민] Good-Night-3rd-Hackathon-Backend 제출

Lauiee opened the pull request:
스프링부트로 구현했습니다.
부족한 실력으로 최대한 열심히 구현했습니다.

Lauiee committed:
최초 커밋

Lauiee committed:
initial commit

Lauiee committed:
initial commit

initial commit

Lauiee committed:
1. 소원등록 API 구현

소원등록 API 구현완료

Lauiee committed:
2. 소원 삭제 구현완료

soft delete방식 소원 삭제 구현 완료

Lauiee committed:
3. 소원 승인상태 변경 API 구현 완료

소원 승인상태 변경 API 구현 완료

Lauiee committed:
3. 승인상태 변경 API 구현 완료

description받는 방식 RequestBody -> RequestParam 으로 변경

Lauiee committed:
4. 소원 단일 조회 API 구현 완료

단일 소원 API구현 완료 일단 승인 상태만 조회 가능하게 만듬

Lauiee committed:
5. 소원 목록 조회 API 구현 완료

테스트코드는 작성 못함

Lauiee committed:
6. 댓글 등록 API 구현 완료

6. 댓글 등록 API 구현 완료

Lauiee committed:
7. 댓글 삭제 API 구현 완료

7. 댓글 삭제 API 구현 완료

Lauiee committed:
8. 댓글 조회 API 구현 완료

8. 댓글 조회 API 구현 완료

Lauiee committed:
fix: 소원 전체 조회 API 수정

Lauiee committed:
9. 검색 API 구현 완료

9. 검색 API 구현 완료

Lauiee committed:
fix: swagger 추가

Lauiee committed:
feat: 추가기능 pending wish 조회 구현

jaeyeon93 commented on src/main/java/com/TecheerTree/myproject/api/controller/CommentsController.java:
```diff
@@ -0,0 +1,47 @@
+    public ResponseEntity<Comments> createComment(@RequestBody CommentSaveRequest commentSaveRequest){
+        Comments comment = commentsService.createComment(commentSaveRequest);
+        return new ResponseEntity<>(comment, HttpStatus.CREATED);
+    }
```
`CommentSaveRequest`는 요청객체라서 service로 전달할때는 도메인객체로 변환하는게 좋습니다.

jaeyeon93 commented on src/main/java/com/TecheerTree/myproject/api/controller/WishController.java:
```diff
@@ -0,0 +1,81 @@
+
+import java.util.List;
+
+@CrossOrigin(origins = "http://localhost:5173") // 프론트엔드 주소를 설정
```
이 옵션 찾아보니 `CORS` 관련 옵션인데 보통은 config로 관리합니다.

jaeyeon93 commented on src/main/java/com/TecheerTree/myproject/api/controller/WishController.java:
```diff
@@ -0,0 +1,81 @@
+    public ResponseEntity<List<Wishes>> getPendingWishes(){
+        List<Wishes> pendingWishes = wishService.getPendingWishes();
+        return new ResponseEntity<>(pendingWishes,HttpStatus.OK);
+    }
```
status 받아서 status로 보류상태 조회하면 재사용성이 높아질거 같아요

jaeyeon93 commented on src/main/java/com/TecheerTree/myproject/api/controller/WishController.java:
```diff
@@ -0,0 +1,81 @@
+    public ResponseEntity<Wishes> createWish(@Valid @RequestBody WishSaveRequest wishSaveRequest){
+        Wishes newWish = wishService.saveWish(wishSaveRequest);
+        return new ResponseEntity<>(newWish, HttpStatus.CREATED);
+    }
```
여기도 request 그대로 넘기기보다는 객체로 변환하는게 좋습니다.

jaeyeon93 commented on src/main/java/com/TecheerTree/myproject/api/controller/WishController.java:
```diff
@@ -0,0 +1,81 @@
+                                                   @RequestParam("description") String description) {
+        Wishes updatedWish = wishService.updateWishStatus(wishId, description);
+        return new ResponseEntity<>(updatedWish, HttpStatus.OK);
+    }
```
저는 pathVariable, requestParam 조합보다는 그냥 requestBody에 다 넣어서 사용하는데, 취향차이이긴한데 고려해보면 좋을거 같아요

Lauiee replied:
다소 단순한 요청은 pathVariable,requestParam조합으로 진행하고,
요청 자체가 좀 복잡해지거나 무거워지면 requestBody를 사용하고 있는데 혹시 그냥 통일하는게 나을까요???

jaeyeon93 commented on src/main/java/com/TecheerTree/myproject/api/repository/CommentsRepository.java:
```diff
@@ -0,0 +1,10 @@
+
+public interface CommentsRepository extends JpaRepository<Comments,Long> {
+    Page<Comments> findByWishIdAndDeletedAtFalse(Long wishId, Pageable pageable);
+}
```
`Comments`가 아니라 `Comment`가 되어야될거 같타요

jaeyeon93 commented on src/main/java/com/TecheerTree/myproject/api/repository/WishRepository.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.List;
+
+@Repository
+public interface WishRepository extends JpaRepository<Wishes, Long> {
```
여기도 Wishes가 아니라 Wish가 되어야될거같아요

jaeyeon93 commented on src/main/java/com/TecheerTree/myproject/api/repository/WishRepository.java:
```diff
@@ -0,0 +1,35 @@
+    Page<Wishes> findAllActive(Pageable pageable);
+
+    @Query("SELECT w FROM Wishes w WHERE w.deleted_at = false AND w.category = :category AND (w.title LIKE %:keyword% OR w.content LIKE %:keyword%)")
+    List<Wishes> findByCategoryAndKeyword(@Param("category") Category category, @Param("keyword") String keyword);
```
or랑 `like %keyword%` 풀스캔 탈 수 있어서 이거 확인해봐야될거에요

jaeyeon93 commented on src/main/java/com/TecheerTree/myproject/api/service/CommentsService.java:
```diff
@@ -0,0 +1,44 @@
+import java.time.LocalDate;
+
+@Service
+@Transactional
```
데이터 변경하느쪽만 Transacitonal 붙이면 될거같아요.

jaeyeon93 commented on src/main/java/com/TecheerTree/myproject/api/service/WishService.java:
```diff
@@ -0,0 +1,136 @@
+
+@Service
+@RequiredArgsConstructor
+@Transactional
```
여기도 데이터 변경하는쪽만 트랜잭션 붙이면될거 같아요.

jaeyeon93 commented on src/main/java/com/TecheerTree/myproject/api/service/WishService.java:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    public Page<WishResponse> getWishesAsResponse(Status status, Pageable pageable) {
+        Page<Wishes> wishes = getWishes(status, pageable);
```
response는 응답으로 나가는 객체인데, service에서 응답객체를 알고있는건 layer 역행이에요. Reponse에서 wish를 받아서 객체 응답으로 나가야합니다.

jaeyeon93 commented on src/main/java/com/TecheerTree/myproject/api/service/WishService.java:
```diff
@@ -0,0 +1,136 @@
+
+    }
+
+    public WishDetailResponse getWish(Long wishId) {
```
여기도 response알고있는거 이상합니다.

jaeyeon93 commented on src/main/java/com/TecheerTree/myproject/api/service/WishService.java:
```diff
@@ -0,0 +1,136 @@
+        WishDetailResponse findWishDto = new WishDetailResponse();
+        findWishDto.setTitle(findWish.getTitle());
+        findWishDto.setContent(findWish.getContent());
+        findWishDto.setCategory(findWish.getCategory().getKoreanName());
```
여기서 response를 만든다고해도 위와 같이 set 방식보다는 `ofCreate`와 같이 정적팩토리 메서드를 사용하는게 좋습니다.

jaeyeon93 commented on src/main/java/com/TecheerTree/myproject/api/service/WishService.java:
```diff
@@ -0,0 +1,136 @@
+        newWish.setStatus(Status.PENDING);
+        newWish.setCreatedDate();
+
+        return wishRepository.save(newWish);
```
set보다는 new로 만들면 불변객체되서 더 좋을거 같아요

jaeyeon93 commented on src/main/java/com/TecheerTree/myproject/domain/dto/response/WishDetailResponse.java:
```diff
@@ -0,0 +1,18 @@
+    }
+
+
+}
```
```
public status WishDetailResponse ofCreate(String title, String contents, String categroy) {
             return new WishDetailResponse(title, contents, category)
}
```
처럼하면 `Setter` 안써도 될거같아요

Lauiee replied:
현재까지 제가 정적 팩토리 메서드로 변경한 코드의 경우 제시해주신 케이스랑 살짝 다르게 인자 하나하나씩 받는것이 아니라,
도메인 객체혹은 DTO를 인자로 받아와서 거기서 getter를 통해 필드값을 가져와 넣어주고 있는데 이렇게 해도 괜찮을까요?
![image](https://github.com/user-attachments/assets/fcb22fb8-3d5d-4400-bf8c-6d98255e2006)
![image](https://github.com/user-attachments/assets/f4bd1d91-305a-42bb-b3cc-7c27efa5f0bd)

jaeyeon93 commented on src/main/java/com/TecheerTree/myproject/domain/entitiy/Wishes.java:
```diff
@@ -0,0 +1,44 @@
+    public Wishes(){
+    }
+
+    public Wishes(String title, String content, Category category, LocalDate createdDate, Status status, boolean deleted_at) {
```
생성자에서 굳이 createDate랑 deleted_at을 받을 필요는없고 내부적으로 LocalDate.now(), false를 넣어주면 됩니다.

jaeyeon93 commented on src/main/java/com/TecheerTree/myproject/domain/entitiy/Wishes.java:
```diff
@@ -0,0 +1,44 @@
+    private String title;
+    private String content;
+    private LocalDate createdDate;
+    private boolean deleted_at;
```
`deleted_at` -> camelCase로 변경해주세요

jaeyeon93 commented on src/main/java/com/TecheerTree/myproject/domain/entitiy/Wishes.java:
```diff
@@ -0,0 +1,44 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "id")
+    private Long wishId;
```
DB의 PK가 될거라 wishId보다는 그냥 id가 더 괜찮은거 같아요. user라고 하면 userId로 대응될텐데, User 테이블의 DB ID인지, 회원의 id인지 헷갈릴거 같아요.

jaeyeon93 commented on src/test/java/com/TecheerTree/myproject/api/service/WishServiceTest.java:
```diff
@@ -0,0 +1,116 @@
+    private WishService wishService;
+
+    @Autowired
+    private WishController wishController;
```
service테스트인데, WishController가 있는게 이상합니다.

Lauiee committed:
refactor: comment생성 컨트롤러 리팩토링

서비스단을 넘기는 요청객체를 도메인 객체로 변환하여 전달

Lauiee committed:
refactor: cors설정을 전역으로 변경

@CrossOrigin으로 컨트롤러 하나에만 적용되던 cors설정을 WebConfig를 생성해 전역으로 적용하게 변경

Lauiee committed:
refactor: Wish생성 컨트롤러 수정

DTO객체 전달 -> 도메인 객체 전달

Lauiee committed:
refactor: service테스트에서 controller 제거

Lauiee committed:
refactor: deleted_at필드 카멜케이스로 변경

Lauiee committed:
refactor: wish객체 생성자 수정

불필요한 인자는 제거하고 내부적으로 값 설정하게 수정

Lauiee committed:
refactor: 데이터 변경 메서드만 @Transactional 적용

Lauiee committed:
refactor: 객체명 단수형으로 변경

Lauiee committed:
refactor: 객체명 단수형으로 변경

빠진 파일들 마저 커밋

Lauiee committed:
refactor: wish객체 id필드명 수정

wishId -> id

Lauiee committed:
refactor: layer역행 수정

불필요했던 메서드 제거 및 서비스 레이어는 도메인 객체만 알도록 수정,
객체 명 및 필드명 수정으로 인해 발생하는 DB관련 오류 수정

Lauiee committed:
refactor: 중복되는 기능의 코드 제거

기존에 만들어져있는 전체 조회 메서드에서 이미 status도 @RequestParam으로 받고 그걸 조건으로 반환해주고 있었어서 불필요한 코드라 판단해서 제거

Lauiee committed:
refactoring: 검색 쿼리에 풀텍스트 서치 적용
